lib/pharos/yaml_file/namespace.rb

```diff
@@ -0,0 +1,26 @@
+# frozen_string_literal: true
+
+module Pharos
+  class YamlFile
+    class Namespace
+      Error = Class.new(Pharos::Error)
+
+      def initialize(path, variables)
+        @path = path
+        variables.each do |key, value|
+          singleton_class.send(:define_method, key) { value }
+        end
+      end
+
+      def inspect
+        super.sub('>', "for='#{@path}'>")
+      end
+
+      def with_binding(&block)
+        yield binding
+      rescue NameError => ex
+        raise Error, "#{ex.message} in file #{@path}"
```

Comment:
Not sure if this exception wrapping is a good idea, this is missing the line number... I think erb itself would be able to handle the `NameError` better if you tell it the `#location=`?

```
undefined local variable or method `asdf' for #<Pharos::YamlFile::Namespace:0x000000000324c568> in file /app/lib/pharos/resources/weave/daemon-set.yml.erb
/app/lib/pharos/yaml_file/namespace.rb:22:in `rescue in with_binding'
/app/lib/pharos/yaml_file/namespace.rb:19:in `with_binding'
/app/lib/pharos/yaml_file.rb:54:in `erb_result'
/app/lib/pharos/yaml_file.rb:50:in `read'
/app/lib/pharos/yaml_file.rb:32:in `load'
/app/lib/pharos/kube.rb:232:in `parse_resource_file'
/app/lib/pharos/kube.rb:129:in `block in apply_stack'
/app/lib/pharos/kube.rb:128:in `each'
/app/lib/pharos/kube.rb:128:in `apply_stack'
/app/lib/pharos/phases/configure_network.rb:50:in `ensure_resources'
/app/lib/pharos/phases/configure_network.rb:25:in `call'
/app/lib/pharos/up_command.rb:131:in `handle_masters'
/app/lib/pharos/up_command.rb:42:in `block in configure'
/app/lib/pharos/up_command.rb:41:in `chdir'
/app/lib/pharos/up_command.rb:41:in `configure'
/app/lib/pharos/up_command.rb:25:in `execute'
/usr/local/bundle/gems/clamp-1.2.1/lib/clamp/command.rb:63:in `run'
/usr/local/bundle/gems/clamp-1.2.1/lib/clamp/subcommand/execution.rb:11:in `execute'
/usr/local/bundle/gems/clamp-1.2.1/lib/clamp/command.rb:63:in `run'
/usr/local/bundle/gems/clamp-1.2.1/lib/clamp/command.rb:132:in `run'
/app/lib/pharos/root_command.rb:14:in `run'
bin/pharos-cluster:12:in `<main>'
```

Reply:
For comparison, removing all the rescue-and-raise bits and modifying the `erb_result` to set the `#location=` instead, the resulting NameError is more useful as it also contains the erb template line number:

```
undefined local variable or method `asdf' for #<Pharos::YamlFile::Namespace:0x0000000001ed8c48>
/app/lib/pharos/resources/weave/daemon-set.yml.erb:30:in `with_binding'
/usr/local/lib/ruby/2.4.0/erb.rb:896:in `eval'
/usr/local/lib/ruby/2.4.0/erb.rb:896:in `result'
/app/lib/pharos/yaml_file.rb:57:in `block in erb_result'
/app/lib/pharos/yaml_file/namespace.rb:20:in `with_binding'
/app/lib/pharos/yaml_file.rb:54:in `erb_result'
/app/lib/pharos/yaml_file.rb:50:in `read'
/app/lib/pharos/yaml_file.rb:32:in `load'
/app/lib/pharos/kube.rb:232:in `parse_resource_file'
/app/lib/pharos/kube.rb:129:in `block in apply_stack'
/app/lib/pharos/kube.rb:128:in `each'
/app/lib/pharos/kube.rb:128:in `apply_stack'
/app/lib/pharos/phases/configure_network.rb:50:in `ensure_resources'
/app/lib/pharos/phases/configure_network.rb:25:in `call'
/app/lib/pharos/up_command.rb:131:in `handle_masters'
/app/lib/pharos/up_command.rb:42:in `block in configure'
/app/lib/pharos/up_command.rb:41:in `chdir'
/app/lib/pharos/up_command.rb:41:in `configure'
/app/lib/pharos/up_command.rb:25:in `execute'
/usr/local/bundle/gems/clamp-1.2.1/lib/clamp/command.rb:63:in `run'
/usr/local/bundle/gems/clamp-1.2.1/lib/clamp/subcommand/execution.rb:11:in `execute'
/usr/local/bundle/gems/clamp-1.2.1/lib/clamp/command.rb:63:in `run'
/usr/local/bundle/gems/clamp-1.2.1/lib/clamp/command.rb:132:in `run'
/app/lib/pharos/root_command.rb:14:in `run'
bin/pharos-cluster:12:in `<main>'
```